File: src/main/java/org/jbb/big/RTreeIndex.java
package org.jbb.big;

import com.google.common.annotations.VisibleForTesting;

import java.io.IOException;
import java.nio.ByteOrder;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Objects;

/**
 * A 1-D R+ tree for storing genomic intervals.
 *
 * TODO: explain that we don't index all intervals when building a tree.
 * TODO: explain that the 1-D R+ tree is simply a range tree build with
 *       using interval union.
 *
 * See tables 14-17 in the Supplementary Data for byte-to-byte details
 * on the R+ tree header and node formats.
 *
 * @author Sergey Zherevchuk
 * @author Sergei Lebedev
 * @since 13/03/15
 */
class RTreeIndex {

  public static RTreeIndex read(final SeekableDataInput s, final long offset)
      throws IOException {
    final Header header = Header.read(s, offset);
    return new RTreeIndex(header);
  }

  @VisibleForTesting
  protected static class Header {

    private static final int MAGIC = 0x2468ace0;

    protected final ByteOrder byteOrder;
    protected final int blockSize;
    protected final long itemCount;
    protected final int startChromIx;
    protected final int startBase;
    protected final int endChromIx;
    protected final int endBase;
    protected final long fileSize;
    protected final int itemsPerSlot;
    protected final long rootOffset; // position of root block of r tree

    Header(final ByteOrder byteOrder, final int blockSize, final long itemCount,
           final int startChromIx, final int startBase, final int endChromIx,
           final int endBase, final long fileSize, final int itemsPerSlot,
           final long rootOffset) {
      this.byteOrder = byteOrder;
      this.blockSize = blockSize;
      this.itemCount = itemCount;
      this.startChromIx = startChromIx;
      this.startBase = startBase;
      this.endChromIx = endChromIx;
      this.endBase = endBase;
      this.fileSize = fileSize;
      this.itemsPerSlot = itemsPerSlot;
      this.rootOffset = rootOffset;
    }

    public static Header read(final SeekableDataInput s, final long offset)
        throws IOException {
      s.seek(offset);
      s.guess(MAGIC);

      final int blockSize = s.readInt();
      final long itemCount = s.readLong();
      final int startChromIx = s.readInt();
      final int startBase = s.readInt();
      final int endChromIx = s.readInt();
      final int endBase = s.readInt();
      final long fileSize = s.readLong();
      final int itemsPerSlot = s.readInt();
      s.readInt(); // skip reserved bytes
      final long rootOffset = s.tell();

      return new Header(s.order(), blockSize, itemCount, startChromIx, startBase,
                        endChromIx, endBase, fileSize, itemsPerSlot, rootOffset);
    }

    public static long write(final SeekableDataOutput writer, final Path chromSizesPath,
                             final Path bedFilePath, final int blockSize,
                             final int itemsPerSlot, final short fieldCount) throws IOException {
      final Hashtable<String, Integer> chromSizesHash = RTreeIndexDetails.bbiChromSizesFromFile(chromSizesPath);
      final wrapObject minDiff = new wrapObject();
      final wrapObject aveSize = new wrapObject(), bedCount = new wrapObject();
      final ArrayList<bbiChromUsage> usageList = RTreeIndexDetails.bbiChromUsageFromBedFile(bedFilePath,
                                                                                      chromSizesHash,
                                                                                      minDiff,
                                                                                      aveSize,
                                                                                      bedCount);

      final int resScales[] = new int[RTreeIndexDetails.bbiMaxZoomLevels];
      final int resSizes[] = new int[RTreeIndexDetails.bbiMaxZoomLevels];
      final int resTryCount = RTreeIndexDetails.bbiCalcResScalesAndSizes(aveSize.toInt(), resScales,
                                                                   resSizes);


      final int blockCount = RTreeIndexDetails.bbiCountSectionsNeeded(usageList, itemsPerSlot);
      final bbiBoundsArray boundsArray[] = new bbiBoundsArray[blockCount];
      for (int i = 0; i < blockCount; i++ ) boundsArray[i] = new bbiBoundsArray();

      final boolean doCompress = false;

      final wrapObject maxBlockSize = new wrapObject();


      RTreeIndexDetails.writeBlocks(usageList, bedFilePath, itemsPerSlot, boundsArray, blockCount,
                                    doCompress, writer,
                                    resTryCount, resScales, resSizes,
                                    bedCount.toInt(), fieldCount, maxBlockSize);

    /* Write out primary data index. */
      final long indexOffset = writer.tell();
      final int itemSize = 24;
      RTreeIndexDetails.cirTreeFileBulkIndexToOpenFile(boundsArray, itemSize, blockCount,
                                                       blockSize, 1,
                                                       indexOffset, writer);

      return indexOffset;
    }
  }

  public final Header header;

  public RTreeIndex(final Header header) {
    this.header = Objects.requireNonNull(header);
  }

  /**
   * Recursively traverses an R+ tree calling {@code consumer} on each
   * block (aka leaf) overlapping a given {@code query}. Note that some
   * of the intervals contained in a block might *not* overlap the
   * {@code query}.
   */
  protected void findOverlappingBlocks(final SeekableDataInput s, final RTreeInterval query,
                                       final UnsafeConsumer<RTreeIndexLeaf> consumer)
      throws IOException {
    final ByteOrder originalOrder = s.order();
    s.order(header.byteOrder);
    try {
      findOverlappingBlocksRecursively(s, query, header.rootOffset, consumer);
    } finally {
      s.order(originalOrder);
    }
  }

  private void findOverlappingBlocksRecursively(final SeekableDataInput s,
                                                final RTreeInterval query,
                                                final long offset,
                                                final UnsafeConsumer<RTreeIndexLeaf> consumer)
      throws IOException {
    // Invariant: a stream is in Header.byteOrder.
    s.seek(offset);

    final boolean isLeaf = s.readBoolean();
    s.readBoolean(); // reserved
    final short childCount = s.readShort();

    if (isLeaf) {
      for (int i = 0; i < childCount; i++) {
        final int startChromIx = s.readInt();
        final int startOffset = s.readInt();
        final int endChromIx = s.readInt();
        final int endOffset = s.readInt();
        final long dataOffset = s.readLong();
        final long dataSize = s.readLong();
        final RTreeIndexNode node
            = RTreeIndexNode.of(startChromIx, startOffset, endChromIx, endOffset,
                                dataOffset);
        if (node.overlaps(query)) {
          final long backup = s.tell();
          consumer.consume(new RTreeIndexLeaf(dataOffset, dataSize));
          s.seek(backup);
        }
      }
    } else {
      final List<RTreeIndexNode> children = new ArrayList<>(childCount);
      for (int i = 0; i < childCount; i++) {
        final int startChromIx = s.readInt();
        final int startBase = s.readInt();
        final int endChromIx = s.readInt();
        final int endBase = s.readInt();
        final long dataOffset = s.readLong();
        final RTreeIndexNode node = RTreeIndexNode.of(
            startChromIx, startBase, endChromIx, endBase, dataOffset);

        // XXX only add overlapping children, because there's no point
        // in storing all of them.
        if (node.overlaps(query)) {
          children.add(node);
        }
      }

      for (final RTreeIndexNode node : children) {
        findOverlappingBlocksRecursively(s, query, node.dataOffset, consumer);
      }
    }
  }
}

File: src/main/kotlin/org/jbb/big/RTreeIndex.kt
package org.jbb.big

import com.google.common.collect.Lists
import com.google.common.primitives.Ints
import com.google.common.primitives.Longs
import com.google.common.primitives.Shorts
import java.io.IOException
import java.nio.ByteOrder
import java.nio.file.Path
import java.util.ArrayList
import java.util.Collections
import java.util.Comparator
import java.util.function.Function

/**
 * A 1-D R+ tree for storing genomic intervals.
 *
 * The tree is built bottom-up by applying union operations to adjacent
 * intervals. The number of intervals combined between levels is
 * defined by the `blockSize` value. For instance for blockSize = 2:
 *
 *    |----------------|  index level 1
 *        /        \
 *    |-----| |--------|  index level 0
 *      / |     /    \
 *    |----|  |--| |---|  leaf  level
 *       |--|
 *
 * The Big format applies a simple trick to optimize tree height.
 * Prior to building the tree the intervals are combined into
 * super-intervals of size `itemsPerSlot`. The R+ tree is then built
 * over these super-intervals.
 *
 * See tables 14-17 in the Supplementary Data for byte-to-byte details
 * on the R+ tree header and node formats.
 */
class RTreeIndex(val header: RTreeIndex.Header) {
    /**
     * Recursively traverses an R+ tree calling `consumer` on each
     * block (aka leaf) overlapping a given `query`. Note that some
     * of the intervals contained in a block might *not* overlap the
     * `query`.
     */
    throws(IOException::class)
    fun findOverlappingBlocks(s: SeekableDataInput, query: ChromosomeInterval,
                              consumer: (RTreeIndexLeaf) -> Unit) {
        val originalOrder = s.order()
        s.order(header.byteOrder)
        try {
            findOverlappingBlocksRecursively(s, query, header.rootOffset, consumer)
        } finally {
            s.order(originalOrder)
        }
    }

    throws(IOException::class)
    private fun findOverlappingBlocksRecursively(input: SeekableDataInput,
                                                 query: ChromosomeInterval,
                                                 offset: Long,
                                                 consumer: (RTreeIndexLeaf) -> Unit) {
        assert(input.order() == header.byteOrder)
        input.seek(offset)

        val isLeaf = input.readBoolean()
        input.readBoolean()  // reserved.
        val childCount = input.readShort().toInt()

        if (isLeaf) {
            for (i in 0 until childCount) {
                val leaf = RTreeIndexLeaf.read(input)
                if (leaf.interval overlaps query) {
                    val backup = input.tell()
                    consumer(leaf)
                    input.seek(backup)
                }
            }
        } else {
            val children = ArrayList<RTreeIndexNode>(childCount.toInt())
            for (i in 0 until childCount) {
                // XXX only add overlapping children, because there's no point
                // in storing all of them.
                val node = RTreeIndexNode.read(input)
                if (node.interval overlaps query) {
                    children.add(node)
                }
            }

            for (node in children) {
                findOverlappingBlocksRecursively(input, query, node.dataOffset, consumer)
            }
        }
    }

    class Header(val byteOrder: ByteOrder, val blockSize: Int, val itemCount: Long,
                 val startChromIx: Int, val startBase: Int,
                 val endChromIx: Int, val endBase: Int,
                 val endDataOffset: Long, val itemsPerSlot: Int, val rootOffset: Long) {
        throws(IOException::class)
        fun write(output: SeekableDataOutput) = with(output) {
            writeInt(MAGIC)
            writeInt(blockSize)
            writeLong(itemCount)
            writeInt(startChromIx)
            writeInt(startBase)
            writeInt(endChromIx)
            writeInt(endBase)
            writeLong(endDataOffset)
            writeInt(itemsPerSlot)
            writeInt(0)  // reserved.
        }

        companion object {
            /** Number of bytes used for this header. */
            val BYTES = Ints.BYTES * 8 + Longs.BYTES * 2
            /** Magic number used for determining [ByteOrder]. */
            private val MAGIC = 0x2468ace0

            throws(IOException::class)
            fun read(input: SeekableDataInput, offset: Long): Header = with(input) {
                seek(offset)
                guess(MAGIC)

                val blockSize = readInt()
                val itemCount = readLong()
                val startChromIx = readInt()
                val startBase = readInt()
                val endChromIx = readInt()
                val endBase = readInt()
                val endDataOffset = readLong()
                val itemsPerSlot = readInt()
                readInt()  // reserved.
                val rootOffset = tell()

                return Header(order(), blockSize, itemCount, startChromIx, startBase,
                              endChromIx, endBase, endDataOffset, itemsPerSlot, rootOffset)
            }
        }
    }

    companion object {
        throws(IOException::class)
        public fun read(input: SeekableDataInput, offset: Long): RTreeIndex {
            return RTreeIndex(Header.read(input, offset))
        }

        throws(IOException::class)
        public fun write(output: SeekableDataOutput, bedPath: Path,
                         blockSize: Int = 1024, itemsPerSlot: Int = 64,
                         compress: Boolean = false): Long {
            require(!compress, "block compression is not supported")

            val bounds = writeBlocks(output, bedPath, itemsPerSlot)

            val dataEndOffset = output.tell()
            val offsets = bounds.map { it.second }.toLongArray()
            // TODO: we can merge adjacent leaves at this point.
            val leaves = bounds.map { it.first }.mapIndexed { i, interval ->
                val endOffset = if (i < offsets.size() - 1) offsets[i + 1] else dataEndOffset
                val size = endOffset - offsets[i]
                RTreeIndexLeaf(interval, offsets[i], size)
            }.toList()

            val wrapper = RTreeIndexWrapper.of(leaves, blockSize)
            val header = Header(output.order(), blockSize, bounds.size().toLong(),
                                wrapper.left.chromIx, wrapper.left.offset,
                                wrapper.right.chromIx, wrapper.right.offset,
                                dataEndOffset, itemsPerSlot, output.tell() + Header.BYTES)
            header.write(output)
            wrapper.write(output, blockSize)

            return dataEndOffset
        }

        throws(IOException::class)
        fun writeBlocks(output: SeekableDataOutput, bedPath: Path,
                         itemsPerSlot: Int): List<Pair<ChromosomeInterval, Long>> {
            var chromId = 0
            val res = Lists.newArrayList<Pair<ChromosomeInterval, Long>>()
            BedFile.read(bedPath).groupBy { it.name }.forEach { entry ->
                val (_name, items) = entry
                Collections.sort(items) { e1, e2 -> Ints.compare(e1.start, e2.start) }

                for (i in 0 until items.size() step itemsPerSlot) {
                    val dataOffset = output.tell()
                    val slotSize = Math.min(items.size() - i, itemsPerSlot)
                    val start = items[i].start
                    var end = items[slotSize - 1].end

                    for (j in 0 until slotSize) {
                        val item = items[i + j]
                        output.writeInt(chromId)
                        output.writeInt(item.start)
                        output.writeInt(item.end)
                        output.writeBytes(item.rest)
                        output.writeByte(0)  // null-terminated.
                    }

                    res.add(ChromosomeInterval(chromId, start, end) to dataOffset)
                }

                chromId++
            }

            return res
        }
    }
}

class RTreeIndexWrapper(private val levels: List<List<Interval>>,
                        private val leaves: List<RTreeIndexLeaf>) {
    private val DUMMY: Offset = Offset(0, 0)

    // TODO: move this to BED summary?
    public val left: Offset
            = if (leaves.isEmpty()) DUMMY else levels.last().first().left
    public val right: Offset
            = if (leaves.isEmpty()) DUMMY else levels.last().last().right

    throws(IOException::class)
    fun write(output: SeekableDataOutput, blockSize: Int) {
        // HEAVY COMPUTER SCIENCE CALCULATION!
        val bytesInNodeHeader = 1 + 1 + Shorts.BYTES
        val bytesInIndexSlot = Ints.BYTES * 4 + Longs.BYTES
        val bytesInIndexBlock = bytesInNodeHeader + blockSize * bytesInIndexSlot
        val bytesInLeafSlot = Ints.BYTES * 4 + Longs.BYTES * 2
        val bytesInLeafBlock = bytesInNodeHeader + blockSize * bytesInLeafSlot

        // Omit root because it's trivial and leaves --- we'll deal
        // with them later.
        levels.subList(1, levels.size() - 1).forEachIndexed { i, level ->
            val bytesInCurrentBlock = bytesInIndexBlock
            val bytesInNextLevelBlock =
                    if (i == levels.size() - 3) bytesInLeafBlock else bytesInCurrentBlock
            var nextChild = output.tell() + bytesInCurrentBlock
            val nodeCount = level.size()
            with(output) {
                writeByte(0)  // isLeaf.
                writeByte(0)  // reserved.
                writeShort(nodeCount)
                for (interval in level) {
                    RTreeIndexNode(interval, nextChild).write(output)
                    nextChild += bytesInNextLevelBlock
                }

                // Write out zeroes for empty slots in node.
                writeByte(0, bytesInIndexSlot * (blockSize - nodeCount))
            }
        }

        with(output) {
            for (i in 0 until leaves.size() step blockSize) {
                val leafCount = Math.min(blockSize, leaves.size() - i)
                writeByte(1)  // isLeaf.
                writeByte(0)  // reserved.
                writeShort(leafCount)
                for (j in 0 until leafCount) {
                    leaves[i + j].write(output)
                }

                // Write out zeroes for empty slots in node.
                writeByte(0, bytesInLeafSlot * (blockSize - leafCount))
            }
        }
    }

    companion object {
        fun of(leaves: List<RTreeIndexLeaf>, blockSize: Int): RTreeIndexWrapper {
            var intervals: List<Interval> = leaves.map { it.interval }
            val levels = arrayListOf(intervals)
            while (intervals.size() > 1) {
                val level = ArrayList<Interval>(intervals.size() / blockSize)
                for (i in 0 until intervals.size() step blockSize) {
                    // |-------|   parent
                    //   /   |
                    //  |-| |-|    links
                    val links = intervals.subList(i, Math.min(intervals.size(), i + blockSize))
                    if (links.size() == 1) {
                        level.addAll(links)
                    } else {
                        level.add(links.reduce(Interval::union))
                    }
                }

                levels.add(level)
                intervals = level
            }

            Collections.reverse(levels)
            return RTreeIndexWrapper(levels, leaves)
        }
    }
}

/**
 * External node aka *leaf* of the chromosome R-tree.
 */
data class RTreeIndexLeaf(public val interval: ChromosomeInterval,
                          public val dataOffset: Long,
                          public val dataSize: Long) {
    fun write(output: SeekableDataOutput) = with(output) {
        writeInt(interval.chromIx)
        writeInt(interval.startOffset)
        writeInt(interval.chromIx)
        writeInt(interval.endOffset)
        writeLong(dataOffset)
        writeLong(dataSize)
    }

    companion object {
        fun read(input: SeekableDataInput): RTreeIndexLeaf = with (input) {
            val startChromIx = readInt()
            val startOffset = readInt()
            val endChromIx = readInt()
            val endOffset = readInt()
            assert(startChromIx == endChromIx)
            val interval = Interval.of(startChromIx, startOffset, endOffset)
            return RTreeIndexLeaf(interval, dataOffset = readLong(),
                                  dataSize = readLong())
        }
    }
}

/**
 * Internal node of the chromosome R-tree.
 */
data class RTreeIndexNode(public val interval: Interval,
                          public val dataOffset: Long) {
    fun write(output: SeekableDataOutput) = with(output) {
        writeInt(interval.left.chromIx)
        writeInt(interval.left.offset)
        writeInt(interval.right.chromIx)
        writeInt(interval.right.offset)
        writeLong(dataOffset)
    }

    companion object {
        fun read(input: SeekableDataInput): RTreeIndexNode = with (input) {
            val startChromIx = readInt()
            val startBase = readInt()
            val endChromIx = readInt()
            val endBase = readInt()
            val interval = Interval.of(startChromIx, startBase, endChromIx, endBase)
            return RTreeIndexNode(interval, dataOffset = readLong())
        }
    }
}